src/_pytest/terminal.py

```diff
@@ -1641,3 +1656,92 @@ def _get_raw_skip_reason(report: TestReport) -> str:
         elif reason == "Skipped":
             reason = ""
         return reason
+
+
+class TerminalProgressPlugin:
+    """Terminal progress reporting plugin using OSC 9;4 ANSI sequences.
+
+    Emits OSC 9;4 sequences to indicate test progress to terminal
+    tabs/windows/etc.
+
+    Not all terminal emulators support this feature.
+
+    Ref: https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC
+    """
+
+    def __init__(self, tr: TerminalReporter) -> None:
+        self._tr = tr
+        self._session: Session | None = None
+        self._has_failures = False
+
+    def _emit_progress(
+        self,
+        state: Literal["remove", "normal", "error", "indeterminate", "paused"],
+        progress: int | None = None,
+    ) -> None:
+        """Emit OSC 9;4 sequence for indicating progress to the terminal.
+
+        :param state:
+            Progress state to set.
+        :param progress:
+            Progress value 0-100. Required for "normal", optional for "error"
+            and "paused", otherwise ignored.
+        """
+        assert progress is None or 0 <= progress <= 100
+
+        # OSC 9;4 sequence: ESC ] 9 ; 4 ; state ; progress ST
+        # ST can be ESC \ or BEL. ESC \ seems better supported.
+        match state:
```

Comment:
Since you say *Changed the code to use `match`, reads more nicely* I'll take the freedom to nitpick on this:

Wouldn't it be simpler to say something like (untested):

```python
state_ids = {
    "remove": 0,
    "normal": 1,
    "error": 2,
    "indeterminate": 3,
    "paused": 4,
}
state_id = state_ids[state]

if progress is None:
    assert state != "normal"
    progress = ""
else:
    assert state not in ["remove", "indeterminate"], state

sequence = f"\x1b]9;4;{state_id};{progress}\x1b\\"
````

Reply:
I find this harder to read, so somewhat prefer the `match` version. Just straight dumb code :)